src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -0,0 +1,282 @@
+                        {
+                            _activeSessionStates.Remove(sessionState);
+                            count = _activeSessionStates.Count;
+                        }
+
+                        if (count == 0)
+                        {
+                            _threadingService.ExecuteSynchronously(() => _projectHotReloadNotificationService.Value.SetHotReloadStateAsync(isInHotReload: false));
+                        }
+                    }, _threadingService);
+
+                    IProjectHotReloadSession projectHotReloadSession = _hotReloadAgent.CreateHotReloadSession(
+                        name: name,
+                        id: _nextUniqueId++,
+                        callback: hotReloadSessionState,
+                        launchProfile: launchProfile,
+                        configuredProject: _configuredProject,
+                        debugLaunchOptions: launchOptions);
+
+                    hotReloadSessionState.Session = projectHotReloadSession;
+                    await projectHotReloadSession.ApplyLaunchVariablesAsync(environmentVariables, default);
+
+                    _pendingSessionState = hotReloadSessionState;
+
+                    return true;
+                }
+                else
+                {
+                    // If startup hooks are not supported then tell the user why Hot Reload isn't available.
+                    string projectName = Path.GetFileNameWithoutExtension(_unconfiguredProject.FullPath);
+
+                    WriteOutputMessage(
+                        new HotReloadLogMessage(
+                            HotReloadVerbosity.Minimal,
+                            Resources.ProjectHotReloadSessionManager_StartupHooksDisabled,
+                            projectName,
+                            null,
+                            HotReloadDiagnosticOutputService.GetProcessId(),
+                            HotReloadDiagnosticErrorLevel.Warning
+                        ),
+                        default);
+                }
+            }
+
+            _pendingSessionState = null;
+            return false;
+        }
+    }
+
+    protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+    {
+        return Task.CompletedTask;
+    }
+
+    protected override Task DisposeCoreAsync(bool initialized)
+    {
+        return _semaphore.ExecuteAsync(DisposeCoreInternalAsync);
+
+        Task DisposeCoreInternalAsync()
+        {
+            foreach (HotReloadSessionState sessionState in _activeSessionStates)
+            {
+                sessionState.Process?.Dispose();
+            }
+
+            _activeSessionStates.Clear();
+
+            return Task.CompletedTask;
+        }
+    }
+
+    private void WriteOutputMessage(HotReloadLogMessage hotReloadLogMessage, CancellationToken cancellationToken) => _hotReloadDiagnosticOutputService.Value.WriteLine(hotReloadLogMessage, cancellationToken);
```

Comment:
Callers are only passing the default cancellation token here. Could remove the parameter, or just set a default on the parameter directly to make the call site simpler.